runtime/major_gc.c

```diff
@@ -148,10 +148,61 @@ static void realloc_gray_vals (void)
   }
 }
 
+#ifdef NATIVE_CODE_AND_NO_NAKED_POINTERS
+
+static int is_pointer_safe (value v, value *p)
+{
+  volatile header_t h;
+  tag_t t;
+
+  Caml_state->checking_pointer_pc = &&on_segfault;
+  h = Hd_val(v);
+  Caml_state->checking_pointer_pc = NULL;
+
+  t = Tag_val(v);
+  if (t == Infix_tag){
+    v -= Infix_offset_val(v);
+    h = Hd_val (v);
+    t = Tag_hd(h);
+  }
+
+  /* For the pointer to be considered safe, either the given pointer is in heap
+   * or the (out of heap) pointer has a black header and its size is < 2 ** 40
+   * words (128 GB). If not, we report a warning. */
+  if (Is_in_heap (v) || (Is_black_hd(h) && Wosize_hd(h) < (1ul << 40)))
+    return 1;
+
+on_segfault:
+  if (Caml_state->checking_pointer_pc == NULL) {
```

Comment:
Maybe I'm missing something obvious, but it's unclear to me why `on_segfault` is here and not on line 186 below.  On a related note, why are we re-testing `checking_ponter_pc`?

Reply:
I needed this to get the optimizer to generate the correct code. I agree that this is quite brittle. 

`checking_pointer_pc` is reset at line 160 if there was no segfault. The segfault handler doesn't reset it. I'll give `sigsetjmp/siglongjmp` a try as it might be the better way to do this.

Reply:
OK, I suspected an issue with optimizations.

There may be a way to turn optimizations off for this function, using GCC/Clang attributes.